compiler/rustc_hir/src/def.rs

```diff
@@ -353,6 +353,14 @@ pub enum Res<Id = hir::HirId> {
     Err,
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, Encodable, Decodable, Hash, Debug)]
+#[derive(HashStable_Generic)]
+pub struct ResImpl {
+    pub def_id: DefId,
+    pub generics_allowed: bool,
+    pub in_trait_ref: bool,
```

Comment:
I'm not sure I understand where this flag is used.

Reply:
We specifically need to deny this:
```rust
impl Trait<Self::A> for () {}
```

But we should not deny this:
```rust
impl Trait<Self> for () {}
```

So we use this flag during astconv to catch cases where we have an associated type with a `Self` in the path, which without this flag would try to elaborate the supertraits of the trait ref, looking for one with the associated type. Since resolving the trait ref would itself require resolving an associated type in the trait ref, we get a cycle error.